File: stockGame-be/src/auth/services/google-oauth.service.ts
import { Injectable, HttpException, HttpStatus } from '@nestjs/common';
import { HttpService } from '@nestjs/axios';
import { ConfigService } from '@nestjs/config';
import { firstValueFrom } from 'rxjs';

export interface GoogleUserInfo {
  id: string;
  email: string;
  verified_email: boolean;
  name: string;
  given_name: string;
  family_name: string;
  picture: string;
  locale: string;
}

@Injectable()
export class GoogleOAuthService {
  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
  ) {}

  getGoogleAuthUrl(): string {
    console.log('🔗 구글 OAuth URL 생성 중...');
    const clientId =
      this.configService.get<string>('google.clientId') ||
      process.env.GOOGLE_CLIENT_ID;
    const redirectUri =
      this.configService.get<string>('google.redirectUri') ||
      process.env.GOOGLE_REDIRECT_URI;

    console.log('📋 구글 OAuth 설정 확인:', {
      clientId: clientId ? '설정됨' : '미설정',
      redirectUri: redirectUri,
    });

    if (!clientId || !redirectUri) {
      console.error('❌ Google OAuth 설정이 누락됨:', {
        clientId: !!clientId,
        redirectUri: !!redirectUri,
      });
      throw new HttpException(
        'Google OAuth 설정이 누락되었습니다.',
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }

    const params = new URLSearchParams({
      client_id: clientId,
      redirect_uri: redirectUri,
      response_type: 'code',
      scope: 'openid email profile',
      access_type: 'offline',
      prompt: 'consent',
    });

    return `https://accounts.google.com/o/oauth2/v2/auth?${params.toString()}`;
  }

  async getAccessToken(authorizationCode: string): Promise<string> {
    console.log('🔑 구글 액세스 토큰 요청 중...');
    const clientId =
      this.configService.get<string>('google.clientId') ||
      process.env.GOOGLE_CLIENT_ID;
    const clientSecret =
      this.configService.get<string>('google.clientSecret') ||
      process.env.GOOGLE_CLIENT_SECRET;
    const redirectUri =
      this.configService.get<string>('google.redirectUri') ||
      process.env.GOOGLE_REDIRECT_URI;

    if (!clientId || !clientSecret || !redirectUri) {
      console.error('❌ Google OAuth 토큰 요청 설정 누락:', {
        clientId: !!clientId,
        clientSecret: !!clientSecret,
        redirectUri: !!redirectUri,
      });
      throw new HttpException(
        'Google OAuth 설정이 누락되었습니다.',
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }

    try {
      const response = await firstValueFrom(
        this.httpService.post('https://oauth2.googleapis.com/token', {
          client_id: clientId,
          client_secret: clientSecret,
          code: authorizationCode,
          grant_type: 'authorization_code',
          redirect_uri: redirectUri,
        }),
      );

      if (response.data.access_token) {
        return response.data.access_token;
      } else {
        throw new HttpException(
          'Google 액세스 토큰을 가져오는데 실패했습니다.',
          HttpStatus.BAD_REQUEST,
        );
      }
    } catch (error) {
      console.error('Google 액세스 토큰 요청 실패:', error);
      throw new HttpException(
        'Google 액세스 토큰을 가져오는데 실패했습니다.',
        HttpStatus.BAD_REQUEST,
      );
    }
  }

  async getUserInfo(accessToken: string): Promise<GoogleUserInfo> {
    try {
      const response = await firstValueFrom(
        this.httpService.get('https://www.googleapis.com/oauth2/v2/userinfo', {
          headers: {
            Authorization: `Bearer ${accessToken}`,
          },
        }),
      );

      return response.data;
    } catch (error) {
      console.error('Google 사용자 정보 요청 실패:', error);
      throw new HttpException(
        'Google 사용자 정보를 가져오는데 실패했습니다.',
        HttpStatus.BAD_REQUEST,
      );
    }
  }
}

File: stockGame-be/src/main/java/kr/knu/getit/stockgame/service/GoogleOAuthService.java
package kr.knu.getit.stockgame.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.Map;
import java.util.Objects;

@Slf4j
@Service
@RequiredArgsConstructor
public class GoogleOAuthService {

    private final RestTemplate restTemplate = new RestTemplate();
    private final ObjectMapper objectMapper = new ObjectMapper();

    @Value("${app.google.client-id}")
    private String clientId;

    @Value("${app.google.client-secret}")
    private String clientSecret;

    @Value("${app.google.redirect-uri}")
    private String redirectUri;

    public String getAuthUrl() {
        return "https://accounts.google.com/o/oauth2/v2/auth"
                + "?client_id=" + clientId
                + "&redirect_uri=" + redirectUri
                + "&response_type=code"
                + "&scope=openid%20email%20profile"
                + "&access_type=offline"
                + "&prompt=consent";
    }

    public String getAccessToken(String code) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        Map<String, Object> body = Map.of(
                "client_id", clientId,
                "client_secret", clientSecret,
                "code", code,
                "grant_type", "authorization_code",
                "redirect_uri", redirectUri
        );

        try {
            String bodyStr = objectMapper.writeValueAsString(body);
            HttpEntity<String> request = new HttpEntity<>(bodyStr, headers);
            ResponseEntity<String> response = restTemplate.postForEntity(
                    "https://oauth2.googleapis.com/token",
                    request,
                    String.class
            );

            if (!response.getStatusCode().is2xxSuccessful() || response.getBody() == null) {
                throw new RuntimeException("Google 액세스 토큰을 가져오는데 실패했습니다.");
            }

            JsonNode node = objectMapper.readTree(Objects.requireNonNull(response.getBody()));
            return node.path("access_token").asText();
        } catch (Exception e) {
            throw new RuntimeException("Google 액세스 토큰 실패", e);
        }
    }

    public GoogleUserInfo getUserInfo(String accessToken) {
        HttpHeaders headers = new HttpHeaders();
        headers.setBearerAuth(Objects.requireNonNull(accessToken));
        HttpEntity<Void> request = new HttpEntity<>(Objects.requireNonNull(headers));
        ResponseEntity<String> response = restTemplate.exchange(
                "https://www.googleapis.com/oauth2/v2/userinfo",
                HttpMethod.GET,
                request,
                String.class
        );

        if (!response.getStatusCode().is2xxSuccessful() || response.getBody() == null) {
            throw new RuntimeException("Google 사용자 정보를 가져오는데 실패했습니다.");
        }

        try {
            JsonNode root = objectMapper.readTree(Objects.requireNonNull(response.getBody()));
            return new GoogleUserInfo(
                    root.path("id").asText(),
                    root.path("email").asText(""),
                    root.path("name").asText(""),
                    root.path("picture").asText("")
            );
        } catch (Exception e) {
            throw new RuntimeException("Google 사용자 정보 파싱 실패", e);
        }
    }

    public record GoogleUserInfo(String id, String email, String name, String picture) {}
}
